### PR TITLE
Esp ota

### PR DESCRIPTION
**Pull request by Gretel5X**
Adds firmware updates via frontend. I integrated it with the latest frontend (conceptional) changes I found in `main` and tweaked it on the way, nothing major. 
The dual-bank management and fail-over is all handled by the esp library
Please check out and really try it on your maschines

### CODE REVIEW

File: main/CMakeLists.txt [hunk 1]
```diff
@@ -37,3 +38,13 @@ spiffs_create_partition_image(website ../webapp/dist FLASH_IN_PROJECT)
 
 # Uncomment to upload STM32 firmware binary saved in ../stm_image/firmware.bin (for testing)
 #spiffs_create_partition_image(stm_ota ../stm_image FLASH_IN_PROJECT)
+find_package(Git)
```
Comment (martinjaeger) [inline main/CMakeLists.txt:42]: Please add newline to separate from previous comment.

File: main/ota.c [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    esp_err_t err;
+    /* update handle : set by esp_ota_begin(), must be freed via esp_ota_end() */
+    esp_ota_handle_t update_handle = 0 ;
+    const esp_partition_t *update_partition = NULL;
+
+    const esp_partition_t *configured = esp_ota_get_boot_partition();
+    const esp_partition_t *running = esp_ota_get_running_partition();
+
+    if (configured != running) {
+        ESP_LOGW(TAG, "Configured OTA boot partition at offset 0x%08x, but running from offset 0x%08x",
```
Comment (martinjaeger) [inline main/ota.c]: Please break lines to stay below 100 characters limit (also for some more lines below).

File: main/ota.c [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    /* deal with all receive packets */
+    bool image_header_was_checked = false;
+    while (binary_file_length - total_received > 0) {
+        data_read = httpd_req_recv(req, ota_write_data, BUFFSIZE);
+        total_received += data_read;
+        if (data_read < 0) {
+            ESP_LOGE(TAG, "Could not receive image");
+            cJSON_AddStringToObject(res, "error", "Could not receive image");
+            return ESP_FAIL;
+        } else if (data_read > 0) {
```
Comment (martinjaeger) [inline main/ota.c]: Please break line after `}` to be compliant with [coding style](https://github.com/LibreSolar/coding-style).

Also applies to a few other `else` statements in this PR.

File: partitions.csv [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
 phy_init, data, phy,     0xf000,  0x1000,
-factory,  app,  factory, 0x10000, 2M,
-website,  data, spiffs,  ,         512K,
-stm_ota,  data, spiffs,  ,         0x20000,
-config,   data, nvs,     ,         0x5000
+ota_0,    app,  ota_0,   0x10000, 1664K,
+ota_1,    app,  ota_1,  ,         1664K,
+website,  data, spiffs, ,         512K,
+stm_ota,  data, spiffs, ,         128K,
+config,   data, nvs,    ,         20K,
```
Comment (martinjaeger) [inline partitions.csv]: Please add one line-break at end of file.

File: webapp/src/components/EspOta.vue [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        .then((res) => {
+          if (target == "old") {
+            this.oldFwVersion = res.data.FirmwareVersion;
+          } else if (target == "new") {
+            this.NewFwVersion = res.data.FirmwareVersion;
+            this.status =
+              "Updated successfully from " +
+              this.oldFwVersion +
+              " --> " +
+              this.NewFwVersion;
```
Comment (martinjaeger) [inline webapp/src/components/EspOta.vue]: This does not seem to work, it keeps the old version:

image

Or is there some special treatment of the firmware image necessary to make it work?

File: webapp/src/components/EspOta.vue [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      this.success = false;
+      this.alert = true;
+    },
+    showSuccess: function () {
+      this.alert = false;
+      this.success = true;
+    },
+  },
+};
+</script>
```
Comment (martinjaeger) [inline webapp/src/components/EspOta.vue]: Please add newline also here.

File: webapp/src/components/EspOta.vue [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        <v-row justify="center">
+            <v-alert
+            v-model="alert"
+            dense
+            outlined
+            type="error"
+            transition="scale-transition"
+            dismissible
+            >{{ status }}</v-alert
+            >
```
Comment (martinjaeger) [inline webapp/src/components/EspOta.vue]: Please use consistent indentation (2 spaces in JavaScript and HTML files, 4 spaces in C/C++ files).

File: webapp/src/components/EspOta.vue [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  props: ['deviceID'],
+  data() {
+    return {
+      uploading: false,
+      file: null,
+      status: null,
+      alert: false,
+      success: false,
+      oldFwVersion: null,
+      NewFwVersion: null,
```
Comment (martinjaeger) [inline webapp/src/components/EspOta.vue]: variable name should start with lower-case character: `newFwVersion`.

File: main/data_nodes.cpp [hunk 9]
```diff
@@ -124,6 +129,9 @@ void data_nodes_init()
     id64 += ((uint64_t)LIBRE_SOLAR_TYPE_ID) << 32;
     uint64_to_base32(id64, device_id, sizeof(device_id));
 
+    const esp_app_desc_t *description = esp_ota_get_app_description();
+    strncpy(firmware_version, description->version, 32);
```
Comment (martinjaeger) [inline main/data_nodes.cpp]: Safer implementation:
```suggestion
    strncpy(firmware_version, description->version, sizeof(firmware_version));
```

**Review by martinjaeger — COMMENTED** [review]
Wow, amazing, works straight away. That's a really neat feature!

If we wanted to update the frontend aswell, we'd probably have to duplicate the process for a different flash partition, correct? (just want to know, not asking to implement it)

My comments are mainly related to coding style. I fixed some inconsistencies in previous code, so I'd like to enforce homogeneous coding style for all upcoming PRs now.

File: webapp/src/components/EspOta.vue (hunk 5 above)
Comment (Gretel5X) [reply webapp/src/components/EspOta.vue]: If you are building the same commit, it will not change

Comment (martinjaeger) [reply webapp/src/components/EspOta.vue]: I created a dummy commit for the new firmware image and it was properly recognized after refreshing the page.

image

Comment (Gretel5X) [reply webapp/src/components/EspOta.vue]: I think the problem is that the request is fired before the esp actually rebooted. I added a 5s delay to wait for reboot.

**Comment by Gretel5X** [comment]
@martinjaeger  I do not have the device with me and were not able to test the changes properly, please build again and try it

**Comment by martinjaeger** [comment]
Cool, thanks. Gonna test it later this week.

File: webapp/src/components/EspOta.vue (hunk 5 above)
Comment (martinjaeger) [reply webapp/src/components/EspOta.vue]: It works now, really cool! Going to merge :)

**Pull request merged**